uniffi_bindgen/src/bindings/gecko_js/webidl.rs

```diff
@@ -2,171 +2,537 @@
-//! arguments. For example, static methods take a `GlobalObject`, methods that
-//! return an `ArrayBuffer` also take a `JSContext`, some return values are
-//! passed via out parameters while others are returned directly, some
-//! arguments map to different C++ types depending on whether they're in or out
-//! parameters, and throwing functions also take an `ErrorResult`.
+//! Although UDL (the UniFFI Definition Language) and WebIDL share the same
+//! syntax, they have completely different semantics. WebIDL distinguishes
+//! between "nullable" (`T?`) and "optional" types (`optional T`), dictionary
+//! members are optional by default, dictionaries are not nullable (but can be
+//! optional, and must have a default value if they are). The UniFFI type
+//! system is much simpler, but can't represent these semantics exactly.
 //!
-//! These conditions and combinations are tricky to express in Askama, so we
-//! handle them in extension traits on the UniFFI types, and keep our templates
-//! clean.
+//!
+//! The Firefox C++ bindings are also peculiar. A C++ function that implements
+//! a WebIDL static or namespace method takes an extra `GlobalObject` argument;
+//! methods that return an `ArrayBuffer` also take a `JSContext`; some return
+//! values are passed via out parameters while others are returned directly;
+//! some WebIDL types map to different C++ types depending on where they're used
+//! (in parameters, out parameters, or dictionary members); and throwing
+//! functions take an extra `ErrorResult` argument.
+//!
+//! https://developer.mozilla.org/en-US/docs/Mozilla/WebIDL_bindings describes
+//! how Firefox WebIDL constructs are reflected in C++. Since UniFFI is
+//! generating both the WebIDL file and the C++ implementation, it must do the
+//! same.
+//!
+//! These combinations of special cases are impossible to express in Askama, so
+//! we have a "shadow type system" that wraps the UniFFI interface types.
+//! Capturing this logic here lets us keep our templates and filters clean.
+
+use crate::interface::{
+    Argument, Constructor, FFIFunction, FFIType, Field, Function, Literal, Method, Object, Record,
+    Type,
+};
+
+/// WebIDL types correspond to UniFFI interface types, s
+#[derive(Clone, Debug)]
+pub enum WebIDLType {
+    /// Wrapped types include integers, floats, Booleans, strings, and enums.
+    /// These wrap the UniFFI type directly, since they have the same names and
+    /// semantics.
+    Wrapped(Type),
+
+    /// Interfaces correspond to UniFFI objects.
+    Interface(String),
+
+    /// Dictionaries correspond to UniFFI records.
+    Dictionary(String),
 
-use crate::interface::{Argument, Constructor, Function, Method, Type};
+    /// `Nullable` and `Optional` both correspond to UniFFI optional types.
+    /// Semantically, "nullable" means "must be passed as an argument or a
+    /// dictionary member, but can be `null`". "Optional" means the argument
+    /// or member can be omitted entirely.
```

Comment:
I don't know where to capture it, but the way I explain this to myself is "most languages don't have both `null` and `undefined`" but the difference is very important to javascript.

Reply:
Oh, that's a great summary, I added that to the comment! (As if that wasn't confusing enough, there's an observable difference between explicitly passing `undefined` and omitting it entirely—`{ foo: undefined }` and `{}`—but thankfully all the specs seem to be moving away from that).